File: coding/src/com/algos/topics/li11_heap/revised/L347_TopKFrequentElements.java
package  com.algos.topics.li11_heap.revised;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class L347_TopKFrequentElements {

	public static void main(String[] args) {
		L347_TopKFrequentElements obj = new L347_TopKFrequentElements();
		System.out.println(Arrays.toString(obj.topKFrequent(new int[] {1,1,1,2,2,3,3,4}, 2)));
	}

	// Problem: https://leetcode.com/problems/top-k-frequent-elements/
	// Idea: This can be solved using heap.
	// Alternative: Hashing and create buckets based on freq. All the nums of same freq would fall in same bucket.
	// Look into buckets from the end as max freq value bucket is at the end.
	public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i=0; i<nums.length; i++){
            map.put(nums[i], map.getOrDefault(nums[i], 0)+1);
        }

        List<Integer>[] buckets = new ArrayList[nums.length+1];	// Top freq can be nums.length if all nums are same
        Arrays.fill(buckets, new ArrayList<>());
		for(Map.Entry<Integer, Integer> entry: map.entrySet()) {
        	List<Integer> bucket = buckets[entry.getValue()];
        	bucket.add(entry.getKey());
        }

        int count = 0;
        int[] ans = new int[k];
        for(int i=buckets.length-1; i>=0; i--) {
			for(int j=0; j<buckets[i].size(); j++) {
				ans[count++] = buckets[i].get(j);
				if(count == k) {
					return ans;
				}
			}
        }
        return ans;
    }
}

File: dsa/concepts/11_heaps/python/l347_top_kfrequent_elements.py
from typing import List
from collections import defaultdict

"""
Problem: https://leetcode.com/problems/top-k-frequent-elements/
Idea: This can be solved using heap.
Alternative: Hashing and create buckets based on freq.
All the nums of same freq would fall in same bucket.
Look into buckets from the end as max freq value bucket is at the end.
"""
def topKFrequent(nums: List[int], k: int) -> List[int]:

    # Frequency map
    freq = defaultdict(int)
    for num in nums:
        freq[num] += 1

    # Buckets where index = frequency
    # Top freq can be len(nums) if all nums are same
    buckets = [[] for _ in range(len(nums) + 1)]
    for num, f in freq.items():
        buckets[f].append(num)

    # Collect result from highest frequency bucket
    result = []
    for f in range(len(buckets) - 1, 0, -1):
        for num in buckets[f]:
            result.append(num)
            if len(result) == k:
                return result

    return result


# ---- one test (similar to Java main) ----
if __name__ == "__main__":
    print(topKFrequent([1, 1, 1, 2, 2, 3, 3, 4], 2))
